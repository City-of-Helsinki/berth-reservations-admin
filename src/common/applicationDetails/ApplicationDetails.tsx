import React from 'react';
import { useTranslation } from 'react-i18next';
import { useLocation } from 'react-router-dom';

import Section from '../section/Section';
import LabelValuePair from '../labelValuePair/LabelValuePair';
import Grid from '../grid/Grid';
import styles from './applicationDetails.module.scss';
import InternalLink from '../internalLink/InternalLink';
import Text from '../text/Text';
import List from '../list/List';
import ListItem from '../list/ListItem';
import { formatDimension, formatWeight, formatDate } from '../utils/format';
import { APPLICATION_STATUS } from '../utils/consonants';
import { ApplicationStatus } from '../../@types/__generated__/globalTypes';
import PrivateCustomerDetails, { PrivateCustomerDetailsProps } from '../privateCustomerDetails/PrivateCustomerDetails';
import OrganizationCustomerDetails, {
  OrganizationCustomerDetailsProps,
} from '../organizationCustomerDetails/OrganizationCustomerDetails';
import Checkbox from '../checkbox/Checkbox';

interface Choice {
  priority: number;
}

interface HarborChoice extends Choice {
  harborName: string;
  harbor: string;
}

interface WinterStorageAreaChoice extends Choice {
  winterStorageAreaName: string;
  winterStorageArea: string;
}

interface Lease {
  berthNum: string | number;
  harborId: string;
  harborName: string;
  id: string;
  pierIdentifier: string;
}

interface BerthSwitch {
  berthNum: string | number;
  harborId: string;
  harborName: string;
  pierIdentifier: string;
  reason: string | null;
}

interface SummaryInformation {
  applicationCode: string;
  acceptBoatingNewsletter: boolean;
  acceptFitnessNews: boolean;
  acceptLibraryNews: boolean;
  acceptOtherCultureNews: boolean;
}

export interface ApplicationDetailsProps {
  accessibilityRequired?: boolean;
  applicant?: PrivateCustomerDetailsProps | OrganizationCustomerDetailsProps;
  berthSwitch?: BerthSwitch | null;
  boatDraught?: number | null;
  boatLength: number;
  boatModel: string;
  boatName: string;
  boatRegistrationNumber: string;
  boatType?: string | null;
  boatWeight?: number | null;
  boatWidth: number;
  createdAt: string;
  customerId?: string;
  handleDeleteLease?: (id: string) => void;
  choices: Array<HarborChoice> | Array<WinterStorageAreaChoice>;
  id: string;
  lease?: Lease | null;
  queue: number | null;
  status: ApplicationStatus;
  summaryInformation?: SummaryInformation;
}

const isHarborChoice = (choice: Choice): choice is HarborChoice => (choice as HarborChoice).harbor !== undefined;

const ApplicationDetails = ({
  id,
  customerId,
  applicant,
  berthSwitch,
  createdAt,
  queue,
  status,
  boatType,
  boatRegistrationNumber,
  boatWidth,
  boatLength,
  boatDraught,
  boatWeight,
  boatName,
  boatModel,
  choices,
  lease,
  handleDeleteLease,
  accessibilityRequired,
  summaryInformation,
}: ApplicationDetailsProps) => {
  const { t, i18n } = useTranslation();
  const routerQuery = new URLSearchParams(useLocation().search);

  return (
    <Grid colsCount={3}>
      <div>
        <Section title={t('applicationList.applicationDetails.application')}>
          <LabelValuePair
            label={t('applicationList.applicationDetails.applicationType')}
            value={
              berthSwitch
                ? t('applicationList.applicationType.switchApplication')
                : t('applicationList.applicationType.newApplication')
            }
          />
          <LabelValuePair
            label={t('applicationList.applicationDetails.receivedDate')}
            value={formatDate(createdAt, i18n.language, true)}
          />
          <LabelValuePair label={t('applicationList.applicationDetails.queueNumber')} value={`${queue || ''}`} />
          <LabelValuePair
            label={t('applicationList.applicationDetails.status')}
            value={t(APPLICATION_STATUS[status]?.label)}
          />
        </Section>
        {berthSwitch && (
          <Section title={t('applicationList.applicationDetails.currentBerth')}>
            <LabelValuePair
              label={t('applicationList.applicationDetails.portAndBerth')}
              value={`${berthSwitch.harborName} ${berthSwitch.pierIdentifier} ${berthSwitch.berthNum}`}
            />
            {berthSwitch.reason !== null && (
              <LabelValuePair label={t('applicationList.applicationDetails.reason')} value={`${berthSwitch.reason}`} />
            )}
          </Section>
        )}
        {applicant &&
          ('organization' in applicant ? (
            <OrganizationCustomerDetails
              {...applicant}
              title={t('applicationList.applicationDetails.applicantCompanyInformation').toUpperCase()}
            />
          ) : (
            <PrivateCustomerDetails
              {...applicant}
              title={t('applicationList.applicationDetails.applicantInformation').toUpperCase()}
            />
          ))}
      </div>
      <div>
        <Section title={t('applicationList.applicationDetails.boatInfo')}>
          <LabelValuePair label={t('applicationList.applicationDetails.boatType')} value={boatType} />
          <LabelValuePair
            label={t('applicationList.applicationDetails.registrationNumber')}
            value={boatRegistrationNumber}
          />
        </Section>
        <Section>
          <LabelValuePair
            label={t('applicationList.applicationDetails.boatWidth')}
            value={formatDimension(boatWidth, i18n.language)}
          />
          <LabelValuePair
            label={t('applicationList.applicationDetails.boatLength')}
            value={formatDimension(boatLength, i18n.language)}
          />
          <LabelValuePair
            label={t('applicationList.applicationDetails.boatDepth')}
            value={formatDimension(boatDraught, i18n.language)}
          />
          <LabelValuePair
            label={t('applicationList.applicationDetails.boatWeight')}
            value={formatWeight(boatWeight, i18n.language)}
          />
        </Section>
        <Section>
          <LabelValuePair label={t('applicationList.applicationDetails.boatName')} value={boatName} />
          <LabelValuePair label={t('applicationList.applicationDetails.boatBrand')} value={boatModel} />
        </Section>
        {summaryInformation && (
          <Section title={t('applicationList.applicationDetails.winterStorageApplicationSummary')}>
            <LabelValuePair
              label={t('applicationList.applicationDetails.applicationCode')}
              value={summaryInformation.applicationCode}
            />
            <br />
            {summaryInformation.acceptBoatingNewsletter && (
              <div>
                <Text>{t('applicationList.applicationDetails.acceptBoatingNewsletter')}</Text>
              </div>
            )}
            {summaryInformation.acceptFitnessNews && (
              <div>
                <Text>{t('applicationList.applicationDetails.acceptFitnessNews')}</Text>
              </div>
            )}
            {summaryInformation.acceptLibraryNews && (
              <div>
                <Text>{t('applicationList.applicationDetails.acceptLibraryNews')}</Text>
              </div>
            )}
            {summaryInformation.acceptOtherCultureNews && (
              <div>
                <Text>{t('applicationList.applicationDetails.acceptOtherCultureNews')}</Text>
              </div>
            )}
          </Section>
        )}
      </div>
      <div>
        {lease ? (
          <Section title={t('applicationList.applicationDetails.connectedLease').toUpperCase()}>
            {[lease.harborName, lease.pierIdentifier, lease.berthNum].filter(Boolean).join(' ')}
            {handleDeleteLease && (
              <button className={styles.deleteButton} onClick={() => handleDeleteLease(lease.id)}>
                <Text color="brand">{t('applicationList.applicationDetails.deleteLease')}</Text>
              </button>
            )}
          </Section>
        ) : (
          choices.length > 0 && (
            <Section
              title={
                isHarborChoice(choices[0])
                  ? t('applicationList.applicationDetails.selectedPorts')
                  : t('applicationList.applicationDetails.selectedWinterStorageAreas')
              }
            >
              <List noBullets>
                {[...choices]
                  .sort((choiceA, choiceB) => choiceA.priority - choiceB.priority)
                  .map((choice, i) => {
                    const target = isHarborChoice(choice) ? choice.harbor : choice.winterStorageArea;
                    const targetName = isHarborChoice(choice) ? choice.harborName : choice.winterStorageAreaName;

                    routerQuery.set(isHarborChoice(choice) ? 'harbor' : 'winter-storage-area', target);

                    return (
                      <ListItem key={i}>
                        <Text>
                          {`${t('applicationList.applicationDetails.choice')} 
                      ${i + 1}: `}
                        </Text>
                        {!!customerId ? (
                          <InternalLink to={`/offer/${id}?${routerQuery}`}>{targetName}</InternalLink>
                        ) : (
                          <Text>{targetName}</Text>
                        )}
                      </ListItem>
                    );
                  })}
              </List>
            </Section>
          )
        )}
        <Section>
          <Checkbox
            id={'accessible'}
            label={t('applicationList.applicationDetails.accessible')}
            checked={accessibilityRequired}
            readOnly
            className={styles.accessibleCheckbox}
          />
        </Section>
      </div>
    </Grid>
  );
};

export default ApplicationDetails;
