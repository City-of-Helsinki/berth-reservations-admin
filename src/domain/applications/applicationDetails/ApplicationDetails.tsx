import React from 'react';
import { useTranslation } from 'react-i18next';

import Section from '../../../common/section/Section';
import LabelValuePair from '../../../common/labelValuePair/LabelValuePair';
import Grid from '../../../common/grid/Grid';
import Checkbox from '../../../common/checkbox/Checkbox';
import styles from './applicationDetails.module.scss';

interface HarborChoice {
  harbor: string;
  priority: number;
}

export interface ApplicationDetailsProps {
  applicationType: string;
  createdAt: string;
  queue: number | null;
  status: string | null;
  boatType?: string | null;
  boatRegistrationNumber: string;
  boatWidth: number;
  boatLength: number;
  boatDraught: number | null;
  boatWeight: number | null;
  boatName: string;
  boatModel: string;
  harborChoices: Array<HarborChoice | null>;
  accessibilityRequired: boolean;
}

const ApplicationDetails: React.SFC<ApplicationDetailsProps> = ({
  applicationType,
  createdAt,
  queue,
  status,
  boatType,
  boatRegistrationNumber,
  boatWidth,
  boatLength,
  boatDraught,
  boatWeight,
  boatName,
  boatModel,
  harborChoices,
  accessibilityRequired,
}) => {
  const { t } = useTranslation();
  const notNull = (choice: HarborChoice | null): choice is HarborChoice =>
    !!choice;

  return (
    <div className={styles.applicationsDetails}>
      <Grid colsCount={3}>
        <Section title={t('applications.applicationDetails.application')}>
          <LabelValuePair
            label={t('applications.applicationDetails.applicationType')}
            value={applicationType}
          />
          <LabelValuePair
            label={t('applications.applicationDetails.receivedDate')}
            value={createdAt}
          />
          <LabelValuePair
            label={t('applications.applicationDetails.queueNumber')}
            value={`${queue || ''}`}
          />
          <LabelValuePair
            label={t('applications.applicationDetails.status')}
            value={status}
          />
        </Section>
        <div>
          <Section title={t('applications.applicationDetails.boatInfo')}>
            <LabelValuePair
              label={t('applications.applicationDetails.boatType')}
              value={boatType}
            />
            <LabelValuePair
              label={t('applications.applicationDetails.registrationNumber')}
              value={boatRegistrationNumber}
            />
          </Section>
          <Section>
            <LabelValuePair
              label={t('applications.applicationDetails.boatWidth')}
              value={`${boatWidth}`}
            />
            <LabelValuePair
              label={t('applications.applicationDetails.boatLength')}
              value={`${boatLength}`}
            />
            <LabelValuePair
              label={t('applications.applicationDetails.boatDepth')}
              value={`${boatDraught || ''}`}
            />
            <LabelValuePair
              label={t('applications.applicationDetails.boatWeight')}
              value={`${boatWeight || ''}`}
            />
          </Section>
          <Section>
            <LabelValuePair
              label={t('applications.applicationDetails.boatName')}
              value={boatName}
            />
            <LabelValuePair
              label={t('applications.applicationDetails.boatBrand')}
              value={boatModel}
            />
          </Section>
        </div>
        <div>
          <Section title={t('applications.applicationDetails.selectedPorts')}>
            {[...harborChoices]
              .filter(notNull)
              .sort((choiceA, choiceB) => choiceA.priority - choiceB.priority)
              .map(({ harbor }, i) => (
                <LabelValuePair
                  key={i}
                  label={`${t('applications.applicationDetails.choice')} ${i +
                    1}`}
                  value={harbor}
                />
              ))}
          </Section>
          <Section>
            <Checkbox
              label={t('applications.applicationDetails.accessible')}
              checked={accessibilityRequired}
              size="large"
              readOnly
            />
          </Section>
        </div>
      </Grid>
    </div>
  );
};

export default ApplicationDetails;
