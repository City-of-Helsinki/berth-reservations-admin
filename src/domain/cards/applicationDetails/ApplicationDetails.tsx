import React from 'react';
import { useTranslation } from 'react-i18next';
import { useLocation } from 'react-router';

import Section from '../../../common/section/Section';
import LabelValuePair from '../../../common/labelValuePair/LabelValuePair';
import Grid from '../../../common/grid/Grid';
import Checkbox from '../../../common/checkbox/Checkbox';
import styles from './applicationDetails.module.scss';
import InternalLink from '../../../common/internalLink/InternalLink';
import Text from '../../../common/text/Text';
import List from '../../../common/list/List';
import ListItem from '../../../common/list/ListItem';
import {
  formatDimension,
  formatWeight,
  formatDate,
} from '../../../common/utils/format';
import { APPLICATION_STATUS } from '../../../common/utils/consonants';
import { ApplicationStatus } from '../../../@types/__generated__/globalTypes';
import PrivateCustomerDetails, {
  PrivateCustomerDetailsProps,
} from '../customerProfileCard/privateCustomerDetails/PrivateCustomerDetails';

interface HarborChoice {
  harborName: string;
  harbor: string;
  priority: number;
}

interface Lease {
  berthNum: string;
  harborId: string;
  harborName: string;
  id: string;
  pierIdentifier: string;
}

interface BerthSwitch {
  berthNum: string;
  harborId: string;
  harborName: string;
  pierIdentifier: string;
  reason: string | null;
}

export interface ApplicationDetailsProps {
  id: string;
  applicant?: PrivateCustomerDetailsProps;
  berthSwitch: BerthSwitch | null;
  createdAt: string;
  queue: number | null;
  status: ApplicationStatus;
  lease?: Lease | null;
  boatType?: string | null;
  boatRegistrationNumber: string;
  boatWidth: number;
  boatLength: number;
  boatDraught: number | null;
  boatWeight: number | null;
  boatName: string;
  boatModel: string;
  harborChoices: Array<HarborChoice | null>;
  accessibilityRequired: boolean;
  handleDeleteLease?: (id: string) => void;
}

const ApplicationDetails: React.SFC<ApplicationDetailsProps> = ({
  id,
  applicant,
  berthSwitch,
  createdAt,
  queue,
  status,
  boatType,
  boatRegistrationNumber,
  boatWidth,
  boatLength,
  boatDraught,
  boatWeight,
  boatName,
  boatModel,
  harborChoices,
  lease,
  handleDeleteLease,
  accessibilityRequired,
}) => {
  const { t, i18n } = useTranslation();
  const notNull = (choice: HarborChoice | null): choice is HarborChoice =>
    !!choice;
  const routerQuery = new URLSearchParams(useLocation().search);

  return (
    <Grid colsCount={3}>
      <div>
        <Section title={t('applications.applicationDetails.application')}>
          <LabelValuePair
            label={t('applications.applicationDetails.applicationType')}
            value={
              berthSwitch !== null
                ? t('applications.applicationType.switchApplication')
                : t('applications.applicationType.newApplication')
            }
          />
          <LabelValuePair
            label={t('applications.applicationDetails.receivedDate')}
            value={formatDate(createdAt, i18n.language, true)}
          />
          <LabelValuePair
            label={t('applications.applicationDetails.queueNumber')}
            value={`${queue || ''}`}
          />
          <LabelValuePair
            label={t('applications.applicationDetails.status')}
            value={t(APPLICATION_STATUS[status]?.label)}
          />
        </Section>
        {berthSwitch !== null && (
          <Section title={t('applications.applicationDetails.currentBerth')}>
            <LabelValuePair
              label={t('applications.applicationDetails.portAndBerth')}
              value={`${berthSwitch.harborName} ${berthSwitch.pierIdentifier} ${berthSwitch.berthNum}`}
            />
            {berthSwitch.reason !== null && (
              <LabelValuePair
                label={t('applications.applicationDetails.reason')}
                value={`${berthSwitch.reason}`}
              />
            )}
          </Section>
        )}
        {applicant && (
          <PrivateCustomerDetails
            {...applicant}
            title={t(
              'applications.applicationDetails.applicantInformation'
            ).toUpperCase()}
          />
        )}
      </div>
      <div>
        <Section title={t('applications.applicationDetails.boatInfo')}>
          <LabelValuePair
            label={t('applications.applicationDetails.boatType')}
            value={boatType}
          />
          <LabelValuePair
            label={t('applications.applicationDetails.registrationNumber')}
            value={boatRegistrationNumber}
          />
        </Section>
        <Section>
          <LabelValuePair
            label={t('applications.applicationDetails.boatWidth')}
            value={formatDimension(boatWidth, i18n.language)}
          />
          <LabelValuePair
            label={t('applications.applicationDetails.boatLength')}
            value={formatDimension(boatLength, i18n.language)}
          />
          <LabelValuePair
            label={t('applications.applicationDetails.boatDepth')}
            value={formatDimension(boatDraught, i18n.language)}
          />
          <LabelValuePair
            label={t('applications.applicationDetails.boatWeight')}
            value={formatWeight(boatWeight, i18n.language)}
          />
        </Section>
        <Section>
          <LabelValuePair
            label={t('applications.applicationDetails.boatName')}
            value={boatName}
          />
          <LabelValuePair
            label={t('applications.applicationDetails.boatBrand')}
            value={boatModel}
          />
        </Section>
      </div>
      <div>
        {lease ? (
          <Section
            title={t(
              'applications.applicationDetails.connectedLease'
            ).toUpperCase()}
          >
            {[lease.harborName, lease.pierIdentifier, lease.berthNum]
              .filter(Boolean)
              .join(' ')}
            {handleDeleteLease && (
              <button
                className={styles.deleteButton}
                onClick={() => handleDeleteLease(lease.id)}
              >
                <Text color="brand">
                  {t('applications.applicationDetails.deleteLease')}
                </Text>
              </button>
            )}
          </Section>
        ) : (
          <Section title={t('applications.applicationDetails.selectedPorts')}>
            <List noBullets>
              {[...harborChoices]
                .filter(notNull)
                .sort((choiceA, choiceB) => choiceA.priority - choiceB.priority)
                .map(({ harborName, harbor }, i) => {
                  routerQuery.set('harbor', harbor);

                  return (
                    <ListItem key={i}>
                      <Text>
                        {`${t('applications.applicationDetails.choice')} 
                      ${i + 1}: `}
                      </Text>
                      <InternalLink to={`/offer/${id}?${routerQuery}`}>
                        {harborName}
                      </InternalLink>
                    </ListItem>
                  );
                })}
            </List>
          </Section>
        )}
        <Section>
          <Checkbox
            label={t('applications.applicationDetails.accessible')}
            checked={accessibilityRequired}
            size="large"
            readOnly
          />
        </Section>
      </div>
    </Grid>
  );
};

export default ApplicationDetails;
