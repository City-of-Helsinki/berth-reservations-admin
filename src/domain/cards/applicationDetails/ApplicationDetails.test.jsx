import React from 'react';
import { shallow } from 'enzyme';
import { HashRouter } from 'react-router-dom';

import ApplicationDetails from './ApplicationDetails';

const mockProps = {
  id: '54321',
  applicationType: 'Vaihtohakemus',
  createdAt: '23.8.2019, klo 21.06',
  queue: 245,
  status: 'Ei kÃ¤sitelty',
  boatType: 'Purjevene / moottoripursi',
  boatRegistrationNumber: 'A 12345',
  boatWidth: 3.2,
  boatLength: 6,
  boatDraught: 0.8,
  boatWeight: 350,
  boatName: 'Cama la Yano',
  boatModel: 'Marine',
  harborChoices: [
    { harbor: '123', priority: 1 },
    { harbor: '321', priority: 3 },
  ],
  accessibilityRequired: true,
};

describe('ApplicationDetails', () => {
  const getWrapper = (props = mockProps) =>
    shallow(
      <HashRouter>
        <ApplicationDetails {...props} />
      </HashRouter>
    );

  it('renders normally', () => {
    const wrapper = getWrapper();

    expect(wrapper.html()).toMatchSnapshot();
  });
});
