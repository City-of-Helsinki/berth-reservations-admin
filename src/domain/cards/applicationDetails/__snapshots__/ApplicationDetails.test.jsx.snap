// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationDetails renders normally 1`] = `
"<div class=\\"grid cols3\\"><article class=\\"section\\"><h4 class=\\"text standard h4 m title\\">HAKEMUS</h4><section class=\\"body\\"><div class=\\"labelValuePair\\"><span class=\\"label standard\\">Hakemustyyppi:</span><span class=\\"value left\\">Uusi hakemus</span></div><div class=\\"labelValuePair\\"><span class=\\"label standard\\">Vastaanotettu:</span><span class=\\"value left\\">23.10.2019 klo 12.15</span></div><div class=\\"labelValuePair\\"><span class=\\"label standard\\">Jononumero:</span><span class=\\"value left\\">245</span></div><div class=\\"labelValuePair\\"><span class=\\"label standard\\">Status:</span><span class=\\"value left\\">Odottaa käsittelyä</span></div></section></article><div><article class=\\"section\\"><h4 class=\\"text standard h4 m title\\">VENEEN TIEDOT</h4><section class=\\"body\\"><div class=\\"labelValuePair\\"><span class=\\"label standard\\">Veneen tyyppi:</span><span class=\\"value left\\">Purjevene / moottoripursi</span></div><div class=\\"labelValuePair\\"><span class=\\"label standard\\">Rekisterinumero:</span><span class=\\"value left\\">A 12345</span></div></section></article><article class=\\"section\\"><section class=\\"body\\"><div class=\\"labelValuePair\\"><span class=\\"label standard\\">Veneen leveys:</span><span class=\\"value left\\">3,2 m</span></div><div class=\\"labelValuePair\\"><span class=\\"label standard\\">Veneen pituus:</span><span class=\\"value left\\">6,0 m</span></div><div class=\\"labelValuePair\\"><span class=\\"label standard\\">Veneen syväys:</span><span class=\\"value left\\">0,8 m</span></div><div class=\\"labelValuePair\\"><span class=\\"label standard\\">Veneen paino:</span><span class=\\"value left\\">350 kg</span></div></section></article><article class=\\"section\\"><section class=\\"body\\"><div class=\\"labelValuePair\\"><span class=\\"label standard\\">Nimi:</span><span class=\\"value left\\">Cama la Yano</span></div><div class=\\"labelValuePair\\"><span class=\\"label standard\\">Merkki:</span><span class=\\"value left\\">Marine</span></div></section></article></div><div><article class=\\"section\\"><h4 class=\\"text standard h4 m title\\">VALITUT SATAMAT</h4><section class=\\"body\\"><ul class=\\"list standard noBullets\\"><li><span class=\\"text standard span\\">Toive 
                      1: </span><a class=\\"internalLink brand\\" href=\\"#/offer/54321?harbor=123\\"></a></li><li><span class=\\"text standard span\\">Toive 
                      2: </span><a class=\\"internalLink brand\\" href=\\"#/offer/54321?harbor=321\\"></a></li></ul></section></article><article class=\\"section\\"><section class=\\"body\\"><span class=\\"text standard span\\">Venesatamissamme ei ole tällä hetkellä esteettömiä venepaikkoja.</span></section></article></div></div>"
`;
